From: Mathieu Trudel-Lapierre <mathieu.trudel-lapierre@canonical.com>
Subject: Reread partition table before wiping when creating new partitions
Bug: https://bugs.freedesktop.org/show_bug.cgi?id=85477
Bug-Debian: https://bugs.debian.org/767457
Bug-Ubuntu: https://launchpad.net/bugs/1460602
Reviewed-By: Marc Deslauriers <mdeslaur@ubuntu.com>
Last-Update: 2016-04-01

---
 src/udiskslinuxpartitiontable.c |    7 +++++++
 1 file changed, 7 insertions(+)

Index: b/src/udiskslinuxpartitiontable.c
===================================================================
--- a/src/udiskslinuxpartitiontable.c
+++ b/src/udiskslinuxpartitiontable.c
@@ -343,6 +343,7 @@ handle_create_partition (UDisksPartition
   uid_t caller_uid;
   gid_t caller_gid;
   gboolean do_wipe = TRUE;
+  gboolean was_partitioned = FALSE;
   GError *error;
 
   error = NULL;
@@ -589,6 +590,10 @@ handle_create_partition (UDisksPartition
   /* this is sometimes needed because parted(8) does not generate the uevent itself */
   udisks_linux_block_object_trigger_uevent (UDISKS_LINUX_BLOCK_OBJECT (object));
 
+  was_partitioned = (udisks_object_peek_partition_table (object) != NULL);
+  if (was_partitioned)
+    udisks_linux_block_object_reread_partition_table (UDISKS_LINUX_BLOCK_OBJECT (object));
+
   /* sit and wait for the partition to show up */
   g_warn_if_fail (wait_data->pos_to_wait_for > 0);
   wait_data->partition_table_object = object;
@@ -644,6 +649,8 @@ handle_create_partition (UDisksPartition
   /* this is sometimes needed because parted(8) does not generate the uevent itself */
   udisks_linux_block_object_trigger_uevent (UDISKS_LINUX_BLOCK_OBJECT (partition_object));
 
+  if (was_partitioned)
+    udisks_linux_block_object_reread_partition_table (UDISKS_LINUX_BLOCK_OBJECT (object));
 
   udisks_partition_table_complete_create_partition (table,
                                                     invocation,
